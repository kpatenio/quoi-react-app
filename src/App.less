// Import LESS colour variables AND theme variables from antd.
// To see the full list of colours: https://ant.design/docs/spec/colors#Base-Color-Palettes
@import '~antd/es/style/themes/default.less';

// import Google fonts as CSS; see https://stackoverflow.com/questions/22877017/how-to-use-google-font-in-less-css-any-other-solution-to-embedded-custom-font
@import (css) url('https://fonts.googleapis.com/css?family=Raleway&display=swap');


// COLORS
@blackBlue: #00152a; // default colour of antd header. Keep in case we want to reuse this.

body {
    font-family: 'Raleway';
    overflow: hidden;
}