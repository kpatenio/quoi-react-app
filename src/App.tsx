import React, { useState } from 'react';
import { Route, Switch, useHistory } from 'react-router-dom';
// import logo from './logo.svg';
import HeaderComponent from './components/HeaderComponent';
import { Layout } from 'antd';
import MainHomepage from './components/MainHomepage';
import FooterComponent from './components/FooterComponent';
import About from './components/About';
import Entry from './components/Entry';

import { useTranslation } from 'react-i18next';

import AppConstants from './AppConstants';

import './App.less';
// import parse from 'html-react-parser'; // if using this, reinstall package

// TODO - styling remove .cit with #{word_id} and children "&nbsp"?

const { Content } = Layout;

const App: React.FC<any> = () => {
    const { t, i18n } = useTranslation();

    const history = useHistory();

    const [dictCode, setDictCode] = useState<string>(AppConstants.DictCode.EF); // To be sent to and used by API.
    const [isEnglishFrenchDict, setIsEnglishFrenchDict] = useState<boolean>(
        true
    ); // by default, set to English-French dictionary

    const handleClickSearch = (word: string | null) => {
        if (word === null || word.trim() === '') {
            console.log('no word');
        } else {
            history.push(`/search/${dictCode}/${word}`);
        }
    };

    const handleClickToggle = () => {
        if (dictCode === AppConstants.DictCode.EF) {
            setDictCode(AppConstants.DictCode.FE);
        } else {
            setDictCode(AppConstants.DictCode.EF);
        }
        setIsEnglishFrenchDict(!isEnglishFrenchDict);
    };

    const toggleText: string = isEnglishFrenchDict
        ? t('toggleLabel_en-fr')
        : t('toggleLabel_fr-en');

    const handleChangeLanguage = () => {
        if (i18n.language === 'en') {
            i18n.changeLanguage('fr');
        } else {
            i18n.changeLanguage('en');
        }
    };

    /* TODO
    Note that antd's ConfigProvider has locale support with it's own translated placeholders. Note that these can usually still be replaced via placeholder prop.
    This depends on the component. For now, let's use i18next! After implementing all translations, we can use ConfigProvider <div className=""></div>

    OR we use i18next provider if load language based on local storage (future concept)
  */

    // SEE HERE: https://fettblog.eu/typescript-react/components/
    // SEE HERE TOO: https://www.robinwieruch.de/react-function-component

    return (
        <div className="App">
            <HeaderComponent
                handleChangeLanguage={handleChangeLanguage}
                toggleText={toggleText}
                onToggle={handleClickToggle}
                onSearch={handleClickSearch}
            />
            <Content className="main" data-testid="main">
                <Switch>
                    <Route
                        path="/"
                        exact
                        render={() => (
                            <MainHomepage
                                handleClickToggle={handleClickToggle}
                                toggleText={toggleText}
                                handleClickSearch={handleClickSearch}
                            />
                        )}
                    />
                    <Route path="/about" exact component={About} />
                    <Route
                        path="/search/:dictCode/:entryId"
                        component={Entry}
                    />
                </Switch>
            </Content>
            <FooterComponent />
        </div>
    );
};

export default App;
