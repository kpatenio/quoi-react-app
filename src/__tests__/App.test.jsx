import React from 'react';
import {
    render,
    fireEvent,
    cleanup
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import App from '../App';

// Use MemoryRouter to maintain react router's context.
// See here: https://reacttraining.com/react-router/web/guides/testing

describe('App', () => {
    describe('initial render', () => {
        afterEach(() => {
            cleanup();
        });
        it('should render App and match snapshots', () => {
            const { container } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should render header', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            getByTestId('header');
        });

        it('should render main content', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            getByTestId('main');
        });

        it('should render footer', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            getByTestId('footer');
        });
    });

    describe('navigating and routing', () => {
        afterEach(() => {
            cleanup();
        });
        it('should render About page if the About nav is selected', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            const aboutButton = getByTestId('header-about');
            fireEvent.click(aboutButton);
            getByTestId('about');
        });

        it('should still display header and footer after changing routes', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            const aboutButton = getByTestId('header-about');
            fireEvent.click(aboutButton);
            getByTestId('header');
            getByTestId('footer');
        });

        it('should display the header searchbar and unmount the homepage searchbar when navigating out of homepage', () => {
            const { queryByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            expect(queryByTestId('searchbar-homepage')).not.toBeNull();
            expect(queryByTestId('searchbar-header')).toBeNull();

            const aboutButton = queryByTestId('header-about');
            fireEvent.click(aboutButton);
            expect(queryByTestId('searchbar-homepage')).toBeNull();
            expect(queryByTestId('searchbar-header')).not.toBeNull();
        });

        it('should navigate to homepage after clicking on the header-logo', () => {
            const { getByTestId } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            const logo = getByTestId('header-logo');
            fireEvent.click(logo);
            getByTestId('homepage');
        });

        // FIXME - this test will fail if server is offline!
        // solution: mock axios!
        it('should navigate to an entry page after entering input via searchbar', () => {
            const { queryByTestId, getByTestId, getByLabelText } = render(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
            const searchbar = getByTestId('searchbar-homepage');
            const searchButton = getByLabelText('icon: search'); // label provided by antd

            fireEvent.input(searchbar, { target: { value: 'test' } });
            expect(searchbar.value).toBe('test');

            fireEvent.click(searchButton);
            // fireEvent.keyDown(searchbar, {key: "Enter", keyCode: 13});
            getByTestId('entry');
            expect(queryByTestId('homepage')).toBeNull();
        });

        it.skip('should render content with successful search', () => {
            // TODO - mock resolve - maybe we should separate getContent
        });
    });
});
