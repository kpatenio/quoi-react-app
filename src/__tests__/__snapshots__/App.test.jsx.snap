// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`just a test should render App 1`] = `
<div
  class="App"
>
  <header
    class="header ant-layout-header"
  >
    <span
      class="logo"
    >
      <a
        class="home"
        href="/"
      >
        « quoi »
      </a>
    </span>
    <nav>
      <ul>
        <li>
          <a
            href="/about"
          >
            about
          </a>
        </li>
      </ul>
    </nav>
    <button
      class="ant-btn languageUIToggle ant-btn-background-ghost"
      data-testid="languageUIToggle"
      type="button"
    >
      <span>
        languageToggle
      </span>
    </button>
  </header>
  <main
    class="main ant-layout-content"
  >
    <hgroup
      class="titleHeader"
    >
      <h1
        id="title"
      >
         
        « quoi »
         
      </h1>
      <h2
        id="slogan"
      >
         
        slogan
         
      </h2>
    </hgroup>
    <div
      class="searchContainer"
    >
      <button
        class="ant-btn dictionaryToggle ant-btn-primary"
        data-testid="dictionaryToggle"
        type="button"
      >
        <span>
          toggleLabel_en-fr
        </span>
        <i
          aria-label="icon: swap"
          class="anticon anticon-swap"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="swap"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M847.9 592H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h605.2L612.9 851c-4.1 5.2-.4 13 6.3 13h72.5c4.9 0 9.5-2.2 12.6-6.1l168.8-214.1c16.5-21 1.6-51.8-25.2-51.8zM872 356H266.8l144.3-183c4.1-5.2.4-13-6.3-13h-72.5c-4.9 0-9.5 2.2-12.6 6.1L150.9 380.2c-16.5 21-1.6 51.8 25.1 51.8h696c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </i>
      </button>
      <span
        class="ant-input-search searchbar ant-input-affix-wrapper ant-input-affix-wrapper-lg"
      >
        <input
          class="ant-input ant-input-lg"
          placeholder="searchbarPlaceholder"
          type="text"
          value=""
        />
        <span
          class="ant-input-suffix"
        >
          <i
            aria-label="icon: search"
            class="anticon anticon-search ant-input-search-icon"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="search"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0 0 11.6 0l43.6-43.5a8.2 8.2 0 0 0 0-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
              />
            </svg>
          </i>
        </span>
      </span>
    </div>
  </main>
  <footer
    class="footer ant-layout-footer"
  >
    <div
      class="github"
    >
      <i
        aria-label="icon: github"
        class="anticon anticon-github github-logo"
      >
        <svg
          aria-hidden="true"
          class=""
          data-icon="github"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0 1 38.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
          />
        </svg>
      </i>
      <a
        href="https://github.com/kpatenio"
        id="github-profile"
        rel="noopener noreferrer"
        target="_blank"
      >
        kpatenio
      </a>
    </div>
    <div
      class="api-footer"
    >
       
      <p
        id="api-message"
      >
         
        footer
        : 
        <span
          id="api-copyright"
        >
          www.Collinsdictionary.com © HarperCollins Publishers Ltd 2014 
        </span>
      </p>
    </div>
  </footer>
</div>
`;
